src/checkpoint_db.rs

```diff
@@ -20,77 +20,90 @@ use exonum::storage::{Database, Patch, Result as StorageResult, Snapshot};
 /// function.
 ///
 /// **Note:** Intended for testing purposes only. Probably inefficient.
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 pub struct CheckpointDb<T> {
-    inner: T,
-    journal: Arc<RwLock<Vec<Patch>>>,
+    inner: Arc<RwLock<CheckpointDbInner<T>>>,
 }
 
-impl<T: Database + Clone> CheckpointDb<T> {
+impl<T: Database> CheckpointDb<T> {
     /// Creates a new checkpointed database that uses the specified `db` as the underlying
     /// data storage.
     pub fn new(db: T) -> Self {
-        CheckpointDb {
-            inner: db,
-            journal: Arc::new(RwLock::new(Vec::new())),
-        }
+        CheckpointDb { inner: Arc::new(RwLock::new(CheckpointDbInner::new(db))) }
+    }
+
+    /// Returns a handler to the database. The handler could be used to roll the database back
+    /// without having the ownership to it.
+    pub fn handler(&self) -> CheckpointDbHandler<T> {
+        CheckpointDbHandler { inner: Arc::clone(&self.inner) }
+    }
+}
+
+impl<T: Database> Database for CheckpointDb<T> {
+    fn snapshot(&self) -> Box<Snapshot> {
+        self.inner
+            .read()
+            .expect("Cannot lock CheckpointDb for snapshot")
+            .snapshot()
+    }
+
+    fn merge(&self, patch: Patch) -> StorageResult<()> {
+        self.inner
+            .write()
+            .expect("Cannot lock CheckpointDb for merge")
+            .merge(patch)
+    }
+
+    fn merge_sync(&self, patch: Patch) -> StorageResult<()> {
+        self.merge(patch)
     }
+}
+
+/// Handler to a checkpointed database.
+#[derive(Debug)]
+pub struct CheckpointDbHandler<T> {
```

Comment:
The only thing I can't understand is this structure. It's exactly the same as `CheckpointDb`. Why do you need it?

Reply:
It is not strictly necessary (it is possible to use the database instead of handler), but it allows one to restrict API to just rollbacks. I've updated docs a bit.